test/params_test.c

```diff
@@ -480,12 +461,15 @@ static int test_case_variant(const OSSL_PARAM *params,
         || !TEST_ptr(BN_native2bn(bignumbin, bignumbin_l, app_p3))
         || !TEST_BN_eq(app_p3, verify_p3)       /* "provider" value */
         || !TEST_str_eq(app_p4, p4_init)        /* "provider" value */
-        || !TEST_size_t_eq(app_p5_l, sizeof(p5_init)) /* "provider" value */
+        || !TEST_ptr(p = OSSL_PARAM_locate(params, "p5"))
+        || !TEST_size_t_eq(p->return_size, sizeof(p5_init)) /* "provider" value */
```

Comment:
For a moment there, I was thinking that no, this is not acceptable! The OSSL_PARAM owner has full control, it shouldn't have to search for those params! Show stopper!
... 
But I realised that the owner is still in control, as it can know the exact location of this parameter and shouldn't need to search for it. I *do* realise it's practical, rather than having to store indexes for the interesting parameters (í.e those you want to check the return size of. Which is better for the sake of being an example?

Reply:
I was more thinking about future changes to the code when an item is added to the array and the indices all go awry.

I think it's more obvious to do the search but it might be possible to cache the pointer as part of the get/set calls -- these come from two different places `static_api_params` and `construct_api_params` so having a statically allocated & initialised pointer isn't viable.

Reply:
> I was more thinking about future changes to the code when an item is added to the array and the indices all go awry.

I don't see why the indices would go awry, and rather see that as a development in the wrong direction.

> I think it's more obvious to do the search but it might be possible to cache the pointer as part of the get/set calls -- these come from two different places `static_api_params` and `construct_api_params` so having a statically allocated & initialised pointer isn't viable.

For `construct_api_params`, it would be easy to save away those positions (either as indices or as direct pointers), but ok, I see the issue with `static_api_params`.

Reply:
Hard coded array index values are just another _magic_ constant.  They should be avoided and are evil :)

Too often, I've experienced having to modify code that used _fixed_ array indices and having to insert something into the middle.  Lots of manually adding one to integers.  Likewise for deleting an entry.  What happens if an extra test case, _p7_, were added in the natural place?  The index to _foo_ would change.  It's not a lot of extra effort but it is error prone.

Reply:
@paulidale, that shouldn't be very hard, you either do this:

``` C
    /* Somewhere, there is a 'OSSL_PARAM *foo_p' */
    foo_p = &params[n];
    params[n++] = OSSL_PARAM_construct_octet_string("foo", &foo, sizeof(foo));
```

... or

``` C
    /* Somewhere, there is a 'size_t foo_i' */
    params[foo_i = n++] = OSSL_PARAM_construct_octet_string("foo", &foo, sizeof(foo));
```

So I'm not sure why you'd need to go for hard-coded values...

Reply:
One of these two is how I'd suggest doing owner side lookups.

What about for `static_api_params`?  Mandating the two arrays stay synchronised is possible but hardly ideal.

Reply:
That is, of course, for `construct_api_params`.  Regarding `static_api_params`, I already said

> but ok, I see the issue with `static_api_params`.

Reply:
So we'll stick with the locate calls?

Reply:
I can drop this one...